xarray/core/dataset.py

```diff
@@ -2782,6 +2782,17 @@ def _rename_all(self, name_dict, dims_dict):
         variables, coord_names = self._rename_vars(name_dict, dims_dict)
         dims = self._rename_dims(dims_dict)
         indexes = self._rename_indexes(name_dict, dims.keys())
+
+        # Variable could be renamed to an existing dimension name
+        # in this case, convert to IndexVariable and set indexes
+        # GH4107
+        for name in coord_names:
```

Comment:
This condition tests whether the new name is an existing coord name and then promotes to IndexVariable. See next comment for a case not caught by this condition.